### PR TITLE
[cozy-lib] Introduce instance-type resource presets (s1/u1/m1)

### PR DESCRIPTION
**Pull request by matthieu-robin**
## Summary

- Replace flat resource presets (`nano`, `micro`, `small`...) with an instance-type naming convention inspired by KubeVirt/cloud providers: `<series>.<size>`
- Three series with different CPU:memory ratios:
  - **s1** (Standard, 1:2) — proxies, caches, lightweight services
  - **u1** (Universal, 1:4) — general purpose databases and messaging
  - **m1** (Memory, 1:8) — memory-intensive workloads (search, analytics)
- Each series provides 8 sizes: `nano`, `micro`, `small`, `medium`, `large`, `xlarge`, `2xlarge`, `4xlarge` (24 presets total)
- Ephemeral storage remains 2Gi for all presets
- Updated default presets in all app/extra packages to use the new naming

## Test plan

- [ ] Verify `helm template` renders correctly for each app with default preset
- [ ] Verify explicit `resources` override still takes priority over preset
- [ ] Run `make unit-tests` to validate existing Helm unit tests
- [ ] Test with a custom preset value to confirm error message on invalid key

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

# Release Notes

* **New Features**
  * Expanded resource sizing presets with three new tier families (s1, u1, m1), offering greater granularity in resource allocation across all managed applications.

* **Changes**
  * Updated default resource presets for applications including ClickHouse, Kafka, Kubernetes, MongoDB, PostgreSQL, Redis, and others to optimize performance and resource utilization.
  * Reorganized resource configuration to support improved sizing flexibility for cloud deployments.

### CODE REVIEW

**Review by kvaps — COMMENTED** [review]
Thanks for working on this!

A couple of things before this can be merged:

**1. Rebase onto main**

The branch is behind main. Since then, OpenSearch was added (#1953), which also uses `resourcesPreset` — it'll need the same preset rename treatment.

**2. Migration script is required**

Existing deployments have HelmReleases with old preset names (`nano`, `small`, `micro`, etc.). After upgrade, the new JSON schemas will reject these values and helm rendering will break.

You need to add a migration script at `packages/core/platform/images/migrations/migrations/36` that:
- Finds all HelmReleases across all namespaces
- Maps old `resourcesPreset` values to new ones (e.g. `small` → `u1.small`, `micro` → `u1.micro`)
- Patches each HelmRelease with `kubectl patch`
- Stamps the new version in the `cozystack-version` ConfigMap

See migration `24` (MongoDB users format conversion) as a good reference — it follows the same pattern of iterating HelmReleases and transforming values with jq.

Also bump `targetVersion` in `packages/core/platform/values.yaml` from `36` to `37`.

**Comment by lexfrei** [comment]
Hi, @matthieu-robin! I'm here to merge this finally. Could you please update this PR?

File: packages/core/platform/images/migrations/migrations/39 [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+echo "Migrating HelmReleases: renaming resourcesPreset values to instance-type format"
+
+# Determine the target series based on HelmRelease name prefix
+get_series() {
+  local name="$1"
+  case "$name" in
+    redis-*|opensearch-*|qdrant-*)
+      echo "m1" ;;
+    harbor-*|vpn-*|tcp-balancer-*|http-cache-*|openbao-*|ingress-*|external-dns-*)
```
Comment (lexfrei) [inline packages/core/platform/images/migrations/migrations/39]: The case patterns `ingress-*` and `external-dns-*` here are dead code. Both `Application`s set `release.prefix: ""` in their cozyrds (`packages/system/ingress-rd/cozyrds/ingress.yaml:13`, `packages/system/external-dns-rd/cozyrds/external-dns.yaml:13`), and `pkg/registry/apps/application/rest.go:237` builds the HelmRelease name as `releaseConfig.Prefix + name`. With prefix empty, the HR name is whatever the user named the CR — `default`, `main`, `edge`, etc. — none of which match `ingress-*` or `external-dns-*`. Users with explicit `resourcesPreset` overrides on these apps fall through to the `u1` default branch and end up with `u1.<size>` (e.g. 2Gi for `u1.micro`), whereas the chart's new default and the apparent intent here is `s1.<size>` (1Gi for `s1.micro`).

Suggested fix: stop matching by HR name regex and key off the kind label that cozystack-api already sets on the HR. The label-based approach is also more durable against future renames — the prefix list here duplicates state that already lives in each `ApplicationDefinition`'s `release.prefix`.

A quick sanity check: of the patterns in this case (`redis-*|opensearch-*|qdrant-*|harbor-*|vpn-*|tcp-balancer-*|http-cache-*|openbao-*|ingress-*|external-dns-*`), the first eight match because their cozyrds set the matching `release.prefix`. The last two never match.

File: packages/core/platform/images/migrations/migrations/39 [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        elif . == "small" then "\($s).small"
+        elif . == "medium" then "\($s).medium"
+        elif . == "large" then "\($s).large"
+        elif . == "xlarge" then "\($s).xlarge"
+        elif . == "2xlarge" then "\($s).2xlarge"
+        else .
+        end;
+
+      walk(
+        if type == "object" and has("resourcesPreset") and (.resourcesPreset | type) == "string" then
```
Comment (lexfrei) [inline packages/core/platform/images/migrations/migrations/39:75]: Observation (non-blocking): the `walk(...)` correctly visits every nested `resourcesPreset`, but the SAME series is applied uniformly across the whole values tree. For apps whose new chart defaults mix series per sub-component, this drifts users with explicit nested overrides:

- clickhouse: top `u1.micro` + `clickhouseKeeper: s1.micro` in `packages/apps/clickhouse/values.yaml`. Migration falls to `u1` default for `clickhouse-*`, so an explicit `clickhouseKeeper.resourcesPreset: micro` becomes `u1.micro` (2Gi) vs the chart's new `s1.micro` (1Gi).
- kafka: top `u1.micro` + `zookeeper: s1.small`. Same drift pattern.
- kubernetes: mixed `u1.small` (apiServer) and `s1.micro` (controllerManager, scheduler, konnectivity); migration uniformly uses `u1`.
- opensearch: top `m1.large` + `dashboards: s1.medium` in `packages/apps/opensearch/values.yaml`. Migration applies `m1`, so an explicit `dashboards.resourcesPreset: medium` becomes `m1.medium` (16Gi) vs chart default `s1.medium` (4Gi).

This affects only users who set `resourcesPreset` on nested fields explicitly — not the common case where the chart default applies — but it silently oversizes sub-components. Switching to label-driven selection (see other comment) plus consulting the chart's per-field default for fields the user did not override would resolve this.

File: packages/apps/rabbitmq/README.md [hunk 3]
```diff
@@ -13,17 +13,17 @@ The service utilizes official RabbitMQ operator. This ensures the reliability an
...
+| Name               | Description                                                                                                                        | Type       | Value     |
+| ------------------ | ---------------------------------------------------------------------------------------------------------------------------------- | ---------- | --------- |
+| `replicas`         | Number of RabbitMQ replicas.                                                                                                       | `int`      | `3`       |
+| `resources`        | Explicit CPU and memory configuration for each RabbitMQ replica. When omitted, the preset defined in `resourcesPreset` is applied. | `object`   | `{}`      |
+| `resources.cpu`    | CPU available to each replica.                                                                                                     | `quantity` | `""`      |
+| `resources.memory` | Memory (RAM) available to each replica.                                                                                            | `quantity` | `""`      |
```
Comment (lexfrei) [inline packages/apps/rabbitmq/README.md:21]: Observation (non-blocking): the preset reference table later in this README (around line 57, outside the diff window) still shows the legacy 7-row layout (`nano | 100m | 128Mi`, etc.); every other regenerated README in this PR (postgres, redis, mariadb, kafka, …) shows the new 24-row s1/u1/m1 table. The `make generate` / `cozyvalues-gen` pass appears to have skipped this package — re-run it for `packages/apps/rabbitmq` and replace the table.

Side note: the values shown there for the legacy presets (`nano | 100m | 128Mi`) don't match what the old `_resourcepresets.tpl` actually defined (`nano | 250m | 128Mi`). The README was drifted from reality before this PR — worth aligning while you regenerate.

**Review by lexfrei — CHANGES_REQUESTED** [review]
The PR is a clean rename overall — `_resourcepresets.tpl` is sound, the 24-preset enum is consistent across all 17 `api/apps/v1alpha1/*/types.go` files, e2e suites are updated, and CodeRabbit's earlier asks (deterministic error message, list parens) and the maintainer's three March blockers (rebase, migration, opensearch) are addressed.

One blocker on the migration script: two of its case-statement prefixes are dead code (inline). Plus two non-blocking inline notes on uniform-series application across nested fields and a stale README table.

This is also a breaking schema change worth flagging in the PR description: the new dynamic CRD enum drops legacy `nano`…`2xlarge` values, so any user-applied `kind: <App>` manifest with a flat preset name will be rejected after upgrade. The migration only rewrites HelmReleases — it does not touch user-managed CR manifests / GitOps repos / Argo apps.

### Resource budget impact (non-blocking, project-wide)

The migration produces sharp memory upsizes for existing deployments because the new s1/u1/m1 ratios decouple memory from the size-name. Old preset memory was roughly 1:1 to CPU; new u1.* is 1:4 and m1.* is 1:8. So when a user-set `resourcesPreset: small` is rewritten to `u1.small` or `m1.small`:

| Old preset (~mem) | After u1 migration | After m1 migration |
|---|---|---|
| nano (128Mi) | u1.nano (1Gi) | m1.nano (2Gi) |
| small (512Mi) | u1.small (4Gi) | m1.small (8Gi) |
| large (2Gi) | u1.large (16Gi) | m1.large (32Gi) |

That is 8x-16x memory across the cluster on platform upgrade. Defensible per the PR's stated philosophy (m1 for memory-intensive workloads), but unannounced — operators with tight node capacity, ResourceQuotas, or cost ceilings may see scheduling failures or surprise spend. Worth pre-flagging in the PR body and the eventual release note so they can right-size before the migration runs.

**Comment by matthieu-robin** [comment]
Thanks for the careful review @lexfrei! All three points addressed in `c3ac8436` (migration) and `115e8eb3` (README). Also pushed `3f84d830` to fix a codegen drift on `api/apps/v1alpha1/kubernetes/zz_generated.deepcopy.go` (missing `Images` `DeepCopy`) that surfaced after the rebase.

### Blocker — `migrations/39:23` (dead `ingress-*` / `external-dns-*` patterns)

Switched to label-driven selection: the migration now reads `apps.cozystack.io/application.kind` from each HelmRelease (set by `cozystack-api` on every HR it creates, see `pkg/apis/apps/v1alpha1/types.go:26`) instead of regexing the HR name. This removes the duplicated `release.prefix` state, fixes the `Ingress` / `ExternalDNS` fall-through to `u1`, and stays correct regardless of the user-chosen CR name. HRs without the label (e.g. created outside `cozystack-api`) keep the previous `u1` fallback.

### Observation — uniform series across nested fields (`migrations/39:56`)

Documented as a known limitation in the script header. A complete fix requires a per-app, per-field map of chart-default series (ClickHouse `clickhouseKeeper`, Kafka `zookeeper`, OpenSearch `dashboards`, Kubernetes `controllerManager`/`scheduler`/`konnectivity`), which felt invasive for a non-blocking item — happy to extend if you'd rather have it covered. The chart defaults still apply for fields users did not override, so the common case is unaffected.

### Observation — stale `packages/apps/rabbitmq/README.md`

Replaced the legacy 7-row preset table with the 24-row `s1`/`u1`/`m1` layout matching the other regenerated READMEs. `cozyvalues-gen` had skipped it (the preset reference table appears to be hand-maintained).

### Resource budget impact

Will add the 8x–16x memory upsize warning to the PR description and the eventual release note as you suggested. The breaking schema-change note (legacy `nano`…`2xlarge` flat enum dropped, user-managed CRs rejected post-upgrade) goes there too.

### CI

Quick checks (codegen, Build, pre-commit, DCO) green on `115e8eb3`. The latest E2E run flaked early on sandbox install (`namespaces "cozy-system" already exists`); the run before it reached the per-app phase and failed exactly the suites you flagged in the budget table — confirming your read. Could you re-trigger E2E?

**Review by lexfrei — APPROVED** [review]
Prior blocker (migration script using dead HR-name patterns) is resolved in c3ac8436 — verified the `apps.cozystack.io/application.kind` label is set on every cozystack-api-managed HelmRelease (`pkg/registry/apps/application/rest.go:199, 531, 1439`), and HRs without the label safely fall through to the `u1` default. The uniform-series limitation is honestly documented in the migration comment. Rabbitmq README is regenerated in 115e8eb3. LGTM. Worth expanding the `release-note` block at release-cut time to flag the upsize on existing deployments (u1 1:4 / m1 1:8 vs old ~1:1).

**Review by lexfrei — CHANGES_REQUESTED** [review]
NOT LGTM (same blocker as last round) — `mergeable_state` is still `dirty`. Recent commits show the author rebased and regenerated schemas/CRDs/types, but `main` has moved further since that rebase, so conflicts are back. Please rebase onto current `main` once more and force-push; I'll re-check the diff after that.

The PR-internal work continues to look clean (10 commits, 117 files, ~+2435/-1020 across the s1/u1/m1 preset rollout); this is purely a branch-hygiene blocker.

**Review by lexfrei — DISMISSED** [review]
LGTM — the merge-conflict blocker from my last round is gone (`mergeable: true`, force-push at 53d7fbd5). The rebase brought in slightly more files (123 vs 117) due to regenerated CRDs/schemas/Go types, which is expected. E2E, Build, pre-commit, and the codegen-drift gate are all green.

Substantively this is the same s1/u1/m1 instance-type preset rollout I LGTM'd on 115e8eb3 before the branch went dirty — no new logic introduced by the rebase, just the regenerated artifacts catching up to current `main`.

**Review by lexfrei — CHANGES_REQUESTED** [review]
Reverting my prior LGTM. The repeated E2E failures on this branch are not flakes — they are the symptom of a behaviour change this PR makes silently, and the same change will hit production users on upgrade.

**Concrete CI evidence** (run 25159603079, attempts 1 and 4, 13h apart, identical signature):

```
Status:  Failed
Reason:  OutOfmemory
Message: Pod was rejected: Node didn't have enough resource: memory, requested: 4429185024, used: 21829788946, capacity: 24533819392
```

Five tests fail at the same stage in both attempts: `foundationdb`, `mongodb`, `redis`, `kubernetes-latest`, `kubernetes-previous`, plus `vminstance` cumulatively. All workloads create themselves successfully but pods are then OOM-rejected by the scheduler. Concurrent runs on `main` and other PRs are green — the e2e infrastructure is fine, this branch is not.

**Why it happens** — the new presets are not behaviour-equivalent to the old ones they replace:

| old size | old mem | s1.* | u1.* | m1.* |
|---|---|---|---|---|
| nano    | 128Mi | 512Mi (4×) | 1Gi (8×)  | 2Gi (16×)  |
| small   | 512Mi | 2Gi (4×)   | 4Gi (8×)  | 8Gi (16×)  |
| medium  | 1Gi   | 4Gi (4×)   | 8Gi (8×)  | 16Gi (16×) |
| large   | 2Gi   | 8Gi (4×)   | 16Gi (8×) | 32Gi (16×) |

CPU also moves at medium and above: `medium` 1→2, `large` 2→4, `xlarge` 4→8, `2xlarge` 8→16. With `memoryAllocationRatio: 1` the memory request equals the limit, so every multiplier above is paid by the scheduler at request time, not just at OOM time.

**Per-app default change** (chart's own `values.yaml`, not user input):

| app | mem | CPU |
|---|---|---|
| redis: nano → m1.nano | 16× | same |
| qdrant: small → m1.small | 16× | same |
| opensearch: large → m1.large | 16× | **2×** |
| mariadb / nats / rabbitmq: nano → u1.nano | 8× | same |
| foundationdb: medium → u1.small | 4× | same |
| mongodb: small → u1.micro | 4× | **0.5×** ↓ |
| postgres: micro → u1.nano | 4× | **0.5×** ↓ |
| clickhouse: small → u1.micro | 4× | **0.5×** ↓ |
| openbao / vpn / tcp-balancer / ingress / external-dns | 4× | same |

Three of these (`mongodb`, `postgres`, `clickhouse`) silently halve CPU on the chart-default path even as they 4× memory. That is a separate decision worth making consciously, not as a side effect of the series choice.

**On upgrade** — migration `39` walks every cozystack-managed HelmRelease and rewrites `resourcesPreset` according to the `apps.cozystack.io/application.kind` label. This runs automatically, with no dry-run, no per-HR diff log, no node-capacity check, no opt-out. Operators with tight node capacity, ResourceQuotas, or cost ceilings will hit one of:

- pods stuck `Pending` (scheduler can't fit the new request),
- ResourceQuota rejection at HelmRelease reconcile,
- autoscaler scale-out → unannounced cost increase,
- on cloud nodes that can't grow: outage.

**On user-managed CRs** — the `enum` for `resourcesPreset` in `values.schema.json` drops the legacy `nano`/`small`/`medium`/etc values entirely. Any GitOps-managed `kind: Postgres` / `kind: Redis` / etc with a flat preset name will be rejected by API validation post-upgrade. Migration 39 only rewrites HRs — it does not touch the user's source manifests. Flux/ArgoCD reconcile loops will fight back against the migration.

**The PR description does not communicate any of this.** The title carries no `!`. The body says "Updated default presets in all app/extra packages to use the new naming" — phrased as a rename. The test plan covers `helm template` rendering, override priority, unit tests, and the invalid-key error message — none of these test upgrade impact, schema compatibility, or capacity. There are no release notes, no upgrade guide, no warning. CodeRabbit's auto-summary line "to optimize performance and resource utilization" does not help.

I flagged a softer version of this in my 28 Apr review (the "resource budget impact, non-blocking" table) and downgraded to LGTM. The CI failures since then have moved this from "may surprise users" to "is currently breaking the e2e cluster, will break tight production clusters on upgrade".

**To unblock**, in priority order:

1. Make the rename behaviour-preserving by default. Map old preset names to the new namespace at their old sizes (so a HelmRelease that had `resourcesPreset: small` ends up at `s1.small = 1 CPU / 512Mi` after migration, not `u1.small = 1 CPU / 4Gi`). The new s1 series is closest to the old shapes already; aligning the size column makes the rename a no-op for resource budget. Larger u1 / m1 sizes remain available as additive choices.
2. OR explicitly mark this as breaking. Add `!` to the title, replace the "Updated default presets" line with a prominent breaking-change block, audit each per-app default change for intentional vs accidental (call out the three CPU halvings), write an upgrade guide that includes a node-capacity check, and add an upgrade smoke test (install on a tight cluster, run migration, verify scheduling).
3. Make migration 39 emit a per-HR before/after log and consider gating it behind a cozystack-config flag so admins run it after a capacity audit.
4. Address the CR-level schema-incompat case explicitly — either keep the old enum values as deprecated aliases that map to the new ones at chart-render time, or document the GitOps-repo update path clearly in the upgrade guide.
